sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -297,19 +253,19 @@ public int getNumShards() {
      * well as sharding information. The policy must return unique and consistent filenames
      * for different windows and panes.
      */
-    public abstract String windowedFilename(WindowedContext c);
+    public abstract ResourceId windowedFilename(
+        ResourceId outputDirectory, WindowedContext c, String extension);
```

Comment:
The FileBasedSink already has to know the CompressionFactory to open the channel. It doesn't make sense to make every policy support passing in the compression factory just to make it know about the extension. This is contextual information, in the sense that every sink needs it and it's not valid across different applications of the policy.